File: frontend/src/api/payment.ts
/**
 * Payment API endpoints (migration WIP)
 * Provides typed wrappers for payment-related APIs.
 *
 * Note: Backend routes may not be enabled yet; callers should handle 404 gracefully.
 */

import { apiClient } from './client'
import type { PaginatedResponse } from '@/types'
import type {
  CheckoutInfoResponse,
  CreateOrderRequest,
  CreateOrderResult,
  PaymentConfig,
  PaymentOrder as CheckoutPaymentOrder,
  SubscriptionPlan,
} from '@/types/payment'

export type PaymentChannel = 'zpay' | 'stripe' | 'admin' | 'activity'
export type UserPaymentProvider = 'zpay' | 'stripe'
export type PaymentPayMethod = 'alipay' | 'wechat'
export type PaymentProviderChannel = `${UserPaymentProvider}_${PaymentPayMethod}`
export type PaymentCreateChannel = PaymentChannel | PaymentPayMethod | PaymentProviderChannel

export interface PaymentChannelOption {
  provider: UserPaymentProvider
  method: PaymentPayMethod
  channel: PaymentProviderChannel
}

export interface PaymentPlan {
  id: string
  name: string
  amount_usd: number
  pay_usd: number
  credits_usd: number
  exchange_rate: number
  discount_rate: number
  available_channels?: PaymentPayMethod[]
  available_channel_options?: PaymentChannelOption[]
  description?: string
  enabled?: boolean
}

export interface PaymentSubscriptionPlan {
  group_id: number
  group_name: string
  description?: string
  platform: string
  price_usd: number
  validity_days: number
  daily_limit_usd?: number | null
  weekly_limit_usd?: number | null
  monthly_limit_usd?: number | null
  exchange_rate: number
  available_channels?: PaymentPayMethod[]
  available_channel_options?: PaymentChannelOption[]
  has_active_subscription?: boolean
}

export type PaymentOrderStatus = 'pending' | 'paid' | 'expired' | 'cancelled' | 'failed' | 'refunded'

export interface PaymentOrder {
  id: number
  order_no: string
  order_type: string
  provider: PaymentChannel
  channel?: string // 实际支付渠道（alipay/wechat）
  biz_group_id?: number
  biz_validity_days?: number
  remark?: string
  amount_cny: number
  amount_usd: number
  total_usd: number
  status: PaymentOrderStatus
  paid_at?: string
  created_at: string
  updated_at: string
  expire_at?: string
}

export async function getPaymentPlans(): Promise<PaymentPlan[]> {
  const { data } = await apiClient.get<PaymentPlan[]>('/payment/plans')
  return data
}

export async function createPaymentOrder(payload: {
  plan_id?: string
  amount_usd?: number
  subscription_group_id?: number
  channel: PaymentCreateChannel
}): Promise<{
  order: PaymentOrder
  pay_url?: string
  qr_url?: string
}> {
  const { data } = await apiClient.post<{
    order: PaymentOrder
    pay_url?: string
    qr_url?: string
  }>('/payment/orders', payload)
  return data
}

export async function getSubscriptionPlans(): Promise<PaymentSubscriptionPlan[]> {
  const { data } = await apiClient.get<PaymentSubscriptionPlan[]>('/payment/subscription-plans')
  return data
}

export async function getMyPaymentOrders(query?: {
  page?: number
  page_size?: number
  status?: PaymentOrderStatus
}): Promise<PaginatedResponse<PaymentOrder>> {
  const { data } = await apiClient.get<PaginatedResponse<PaymentOrder>>('/payment/orders', {
    params: query
  })
  return data
}

export async function getPaymentOrder(orderNo: string): Promise<PaymentOrder> {
  const { data } = await apiClient.get<PaymentOrder>(`/payment/orders/${orderNo}`)
  return data
}

export async function cancelPaymentOrder(orderNo: string): Promise<{ message: string }> {
  const { data } = await apiClient.post<{ message: string }>(`/payment/orders/${orderNo}/cancel`)
  return data
}

// Compatibility wrappers for the richer payment UI introduced on the tag side.
async function getConfig(): Promise<{ data: PaymentConfig }> {
  const { data } = await apiClient.get<PaymentConfig>('/payment/checkout-info')
  return { data }
}

async function getCheckoutInfo(): Promise<{ data: CheckoutInfoResponse }> {
  const { data } = await apiClient.get<CheckoutInfoResponse>('/payment/checkout-info')
  return { data }
}

async function getPlans(): Promise<{ data: SubscriptionPlan[] }> {
  const { data } = await apiClient.get<SubscriptionPlan[]>('/payment/subscription-plans')
  return { data }
}

async function createOrder(payload: CreateOrderRequest): Promise<{ data: CreateOrderResult }> {
  const { data } = await apiClient.post<CreateOrderResult>('/payment/orders', payload)
  return { data }
}

async function getOrder(orderId: number | string): Promise<{ data: CheckoutPaymentOrder }> {
  const { data } = await apiClient.get<CheckoutPaymentOrder>(`/payment/orders/${orderId}`)
  return { data }
}

async function getMyOrders(query?: {
  page?: number
  page_size?: number
  status?: string
}): Promise<{ data: PaginatedResponse<CheckoutPaymentOrder> }> {
  const { data } = await apiClient.get<PaginatedResponse<CheckoutPaymentOrder>>('/payment/orders/my', {
    params: query,
  })
  return { data }
}

async function cancelOrder(orderId: number | string): Promise<{ data: { message: string } }> {
  const { data } = await apiClient.post<{ message: string }>(`/payment/orders/${orderId}/cancel`)
  return { data }
}

async function requestRefund(
  orderId: number | string,
  payload: { reason: string },
): Promise<{ data: { message: string } }> {
  const { data } = await apiClient.post<{ message: string }>(`/payment/orders/${orderId}/refund`, payload)
  return { data }
}

async function getRefundEligibleProviders(): Promise<{ data: { provider_instance_ids: string[] } }> {
  const { data } = await apiClient.get<{ provider_instance_ids: string[] }>('/payment/providers/refund-eligible')
  return { data }
}

async function verifyOrderPublic(outTradeNo: string): Promise<{ data: CheckoutPaymentOrder }> {
  const { data } = await apiClient.post<CheckoutPaymentOrder>('/payment/public/orders/verify', {
    out_trade_no: outTradeNo,
  })
  return { data }
}

async function resolveOrderPublicByResumeToken(resumeToken: string): Promise<{ data: CheckoutPaymentOrder }> {
  const { data } = await apiClient.post<CheckoutPaymentOrder>('/payment/public/orders/resolve', {
    resume_token: resumeToken,
  })
  return { data }
}

export const paymentAPI = {
  getPaymentPlans,
  getSubscriptionPlans,
  createPaymentOrder,
  getMyPaymentOrders,
  getPaymentOrder,
  cancelPaymentOrder,
  getConfig,
  getCheckoutInfo,
  getPlans,
  createOrder,
  getOrder,
  getMyOrders,
  cancelOrder,
  requestRefund,
  getRefundEligibleProviders,
  verifyOrderPublic,
  resolveOrderPublicByResumeToken,
}

export default paymentAPI

File: backend/internal/handler/dto/payment.go
package dto

import (
	"strings"
	"time"

	"github.com/Wei-Shaw/sub2api/internal/service"
)

// PaymentPlan represents a frontend-friendly payment package definition.
type PaymentPlan struct {
	ID                string                 `json:"id"`
	Name              string                 `json:"name"`
	AmountUSD         float64                `json:"amount_usd"`
	PayUSD            float64                `json:"pay_usd"`
	CreditsUSD        float64                `json:"credits_usd"`
	ExchangeRate      float64                `json:"exchange_rate"`
	DiscountRate      float64                `json:"discount_rate"`
	Enabled           bool                   `json:"enabled"`
	AvailableChannels []string               `json:"available_channels,omitempty"`
	AvailableOptions  []PaymentChannelOption `json:"available_channel_options,omitempty"`
}

type PaymentSubscriptionPlan struct {
	GroupID               int64                  `json:"group_id"`
	GroupName             string                 `json:"group_name"`
	Description           string                 `json:"description,omitempty"`
	Platform              string                 `json:"platform"`
	PriceUSD              float64                `json:"price_usd"`
	ValidityDays          int                    `json:"validity_days"`
	DailyLimitUSD         *float64               `json:"daily_limit_usd,omitempty"`
	WeeklyLimitUSD        *float64               `json:"weekly_limit_usd,omitempty"`
	MonthlyLimitUSD       *float64               `json:"monthly_limit_usd,omitempty"`
	ExchangeRate          float64                `json:"exchange_rate"`
	AvailableChannels     []string               `json:"available_channels,omitempty"`
	AvailableOptions      []PaymentChannelOption `json:"available_channel_options,omitempty"`
	HasActiveSubscription bool                   `json:"has_active_subscription"`
}

type PaymentChannelOption struct {
	Provider string `json:"provider"`
	Method   string `json:"method"`
	Channel  string `json:"channel"`
}

// PaymentOrder represents a payment order for API responses.
type PaymentOrder struct {
	ID        int64   `json:"id"`
	OrderNo   string  `json:"order_no"`
	OrderType string  `json:"order_type"`
	TradeNo   *string `json:"trade_no,omitempty"`

	UserID    int64  `json:"user_id"`
	Username  string `json:"username"`
	UserEmail string `json:"user_email,omitempty"` // admin list/export convenience

	Remark       string  `json:"remark,omitempty"`
	AmountCNY    float64 `json:"amount_cny"`
	AmountUSD    float64 `json:"amount_usd"`
	BonusUSD     float64 `json:"bonus_usd"`
	TotalUSD     float64 `json:"total_usd"`
	ExchangeRate float64 `json:"exchange_rate"`
	DiscountRate float64 `json:"discount_rate"`

	Provider        string `json:"provider"`
	Channel         string `json:"channel"`
	PaymentMethod   string `json:"payment_method"`
	PaymentURL      string `json:"payment_url,omitempty"`
	BizGroupID      *int64 `json:"biz_group_id,omitempty"`
	BizValidityDays *int   `json:"biz_validity_days,omitempty"`

	Status   string     `json:"status"`
	PaidAt   *time.Time `json:"paid_at,omitempty"`
	ExpireAt time.Time  `json:"expire_at"`

	PromotionTier *int `json:"promotion_tier,omitempty"`
	PromotionUsed bool `json:"promotion_used"`

	CallbackData string     `json:"callback_data,omitempty"`
	CallbackAt   *time.Time `json:"callback_at,omitempty"`

	ClientIP  string `json:"client_ip,omitempty"`
	UserAgent string `json:"user_agent,omitempty"`

	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

func PaymentOrderFromService(o *service.PaymentOrder) *PaymentOrder {
	if o == nil {
		return nil
	}
	orderType := "online_recharge"
	if strings.EqualFold(o.Provider, "admin") {
		orderType = "admin_recharge"
	} else if strings.EqualFold(o.Provider, "activity") {
		orderType = "activity_recharge"
	} else if o.IsSubscriptionPurchase() {
		orderType = "subscription_purchase"
	}
	return &PaymentOrder{
		ID:              o.ID,
		OrderNo:         o.OrderNo,
		OrderType:       orderType,
		TradeNo:         o.TradeNo,
		UserID:          o.UserID,
		Username:        o.Username,
		Remark:          o.Remark,
		AmountCNY:       o.AmountCNY,
		AmountUSD:       o.AmountUSD,
		BonusUSD:        o.BonusUSD,
		TotalUSD:        o.TotalUSD,
		ExchangeRate:    o.ExchangeRate,
		DiscountRate:    o.DiscountRate,
		Provider:        o.Provider,
		Channel:         o.Channel,
		PaymentMethod:   o.PaymentMethod,
		PaymentURL:      o.PaymentURL,
		BizGroupID:      o.BizGroupID,
		BizValidityDays: o.BizValidityDays,
		Status:          o.Status,
		PaidAt:          o.PaidAt,
		ExpireAt:        o.ExpireAt,
		PromotionTier:   o.PromotionTier,
		PromotionUsed:   o.PromotionUsed,
		CallbackData:    o.CallbackData,
		CallbackAt:      o.CallbackAt,
		ClientIP:        o.ClientIP,
		UserAgent:       o.UserAgent,
		CreatedAt:       o.CreatedAt,
		UpdatedAt:       o.UpdatedAt,
	}
}
